runtime/moonbase/tests/runtime_apis.rs

```diff
@@ -298,6 +300,95 @@ fn txpool_runtime_api_extrinsic_filter() {
 	});
 }
 
+#[test]
+fn can_author_when_selected_is_empty() {
+	ExtBuilder::default()
+		.with_balances(vec![
+			(AccountId::from(ALICE), 20_000_000 * UNIT),
+			(AccountId::from(BOB), 10_000_000 * UNIT),
+		])
+		.with_collators(vec![(AccountId::from(ALICE), 2_000_000 * UNIT)])
+		.with_mappings(vec![(
+			NimbusId::from_slice(&ALICE_NIMBUS).unwrap(),
+			AccountId::from(ALICE),
+		)])
+		.build()
+		.execute_with(|| {
+			set_parachain_inherent_data();
+			run_to_block(2, Some(NimbusId::from_slice(&ALICE_NIMBUS).unwrap()));
+
+			assert_eq!(ParachainStaking::candidate_pool().0.len(), 1);
+
+			let slot_number = 0;
+			let parent = Header {
+				digest: Default::default(),
+				extrinsics_root: Default::default(),
+				number: Default::default(),
+				parent_hash: Default::default(),
+				state_root: Default::default(),
+			};
+
+			// Base case: ALICE can author blocks when she is the only candidate
+			let can_author_block = Runtime::can_author(
+				NimbusId::from_slice(&ALICE_NIMBUS).unwrap(),
+				slot_number,
+				&parent,
+			);
+
+			assert!(can_author_block);
+
+			// Remove ALICE from candidate pool, leaving the candidate_pool empty
+			assert_ok!(ParachainStaking::go_offline(origin_of(AccountId::from(
+				ALICE
+			))));
+
+			// Need to fast forward to right before the next session, which is when selected candidates
+			// will be updated. We want to test the creation of the first block of the next session.
+			run_to_block(1799, Some(NimbusId::from_slice(&ALICE_NIMBUS).unwrap()));
+
+			assert_eq!(ParachainStaking::candidate_pool().0.len(), 0);
+
+			let slot_number = 0;
+			let parent = Header {
+				digest: Default::default(),
+				extrinsics_root: Default::default(),
+				number: 1799,
+				parent_hash: Default::default(),
+				state_root: Default::default(),
+			};
+
+			let can_author_block = Runtime::can_author(
+				NimbusId::from_slice(&ALICE_NIMBUS).unwrap(),
+				slot_number,
+				&parent,
+			);
+
+			assert!(can_author_block);
+
+			// Check that it works as expected after session update
+			run_to_block(1800, Some(NimbusId::from_slice(&ALICE_NIMBUS).unwrap()));
+
+			assert_eq!(ParachainStaking::candidate_pool().0.len(), 0);
+
+			let slot_number = 0;
+			let parent = Header {
+				digest: Default::default(),
+				extrinsics_root: Default::default(),
+				number: 1800,
```

Comment:
We should test this on other blocks (e.g. `1801`) which aren't on round-change boundaries.

Reply:
Oh, maybe I'm misunderstanding the problem. Is this only a problem on round change boundaries?

Reply:
Yes, this problem only happens on round change. If at least one collator uses `--force-authorship` and creates the first block of the new round, then the rest can keep producing blocks.